CefGlue.Common.Shared/Serialization/Deserializer.cs

```diff
@@ -0,0 +1,275 @@
+                        // they can only be stored in an object after all of the properties/fields that were set in the deserialization
+                        var stateHoldingArrayOrObject = currentState == ListWrapperMarker ? previousStateIfEndArray : currentState;
+                        if (stateHoldingArrayOrObject == null)
+                        {
+                            throw new InvalidOperationException("A ListWrapperMarker must always be preceeded by a state that holds an array.");
+                        }
+                        var parentState = state.Peek();
+                        parentState.SetValue(stateHoldingArrayOrObject.ObjectHolder);
+                        break;
+                }
+
+                previousStateIfEndArray = reader.TokenType == JsonTokenType.EndArray ? currentState : null;
+            } while (reader.Read());
+
+            if (state.Count > 1 && state.Pop().ObjectHolder != root)
+            {
+                throw new InvalidOperationException("Invalid json format - missing enclosing EndArray or EndObject token(s).");
+            }
+
+            return root.First();
+        }
+
+        private static object ReadComplexObject(ref Utf8JsonReader reader, Stack<IDeserializerState> state, bool isRootObjectHolder, ParametersTypes parametersTypes, IDictionary<string, ReferenceInfo> referencesMap)
+        {
+            var tempReader = reader;
+            object obj;
+            IDeserializerState newState = null;
+            JsonTypeInfo newTypeInfo = null;
+
+            tempReader.ReadToken(JsonTokenType.StartObject);
+
+            if (tempReader.TokenType == JsonTokenType.EndObject)
+            {
+                // empty object
+                obj = new object();
+            }
+            else
+            {
+                // peek first field name
+                var propName = tempReader.ReadPropertyName();
+                IDeserializerState currentState;
+                switch (propName)
+                {
+                    case JsonAttributes.Ref:
+                        reader.ReadToken(JsonTokenType.StartObject);
+                        reader.ReadPropertyName();  // skip the $ref
+                        var refId = reader.GetString();
+                        var refInfo = referencesMap[refId];
+                        obj = refInfo.Object;
+                        newTypeInfo = refInfo.TypeInfo;
+                        newState = CreateNewDeserializerState(obj, newTypeInfo);
+                        break;
+
+                    case JsonAttributes.Id:
+                        reader.ReadToken(JsonTokenType.StartObject);
+                        reader.ReadPropertyName(); // skip the $id
+                        var id = tempReader.ReadString();
+                        propName = tempReader.ReadPropertyName();
+                        if (propName == JsonAttributes.Values)
```

Comment:
avoid reusing variables
```suggestion
                        if (tempReader.ReadPropertyName() == JsonAttributes.Values)
```